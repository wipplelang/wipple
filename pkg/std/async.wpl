[[no-std]]

use "math.wpl"

[help "Wait for a function to be called before returning."]
[help-group "Control Flow"]
with-continuation :: A => ((A -> ()) -> ()) -> A
with-continuation : f -> external "runtime" "with-continuation" f

[help "Create a group of tasks to be executed at the same time. This function returns once all of the tasks have finished."]
[help-group "Control Flow"]
with-task-group :: (Task-Group -> ()) -> ()
with-task-group : f -> external "runtime" "with-task-group" f

task' :: (() -> ()) -> Task-Group -> ()
task' : f group -> external "runtime" "task" group f

task : syntax {
	task 'body -> task' (() -> 'body)
}

in-background' :: (() -> ()) -> ()
in-background' : f -> external "runtime" "in-background" f

in-background : syntax {
	in-background 'body -> in-background' (() -> 'body)
}

Duration : type {
	ms :: Natural
}

[help "Create a duration in milliseconds."]
[help-group "Control Flow"]
ms :: Natural -> Duration
ms : ms -> Duration { ms }

[help "Create a duration in seconds."]
[help-group "Control Flow"]
secs :: Natural -> Duration
secs : secs -> (secs * 1000) . ms

[help "Wait for the specified duration to elapse before continuing the program."]
[help-group "Control Flow"]
delay :: Duration -> ()
delay : { ms } -> external "runtime" "delay" ms
