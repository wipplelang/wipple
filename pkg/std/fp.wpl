[[no-std]]

[help "Returns its input unmodified."]
it :: A => A -> A
it : x -> x

[help "Ignores its second input."]
just :: A => A -> _ -> A
just : x -> _ -> x

. : dot operator (x f ~> f x)

Map : A B Input Output => trait ((A -> B) -> Input -> Output)
| : dot operator (x f ~> Map f x)
