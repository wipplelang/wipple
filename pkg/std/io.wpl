[[no-std]]

use "fp.wpl"
use "list.wpl"
use "logic.wpl"
use "util.wpl"
use "iterate.wpl"

[language "show"]
[help "Implement this trait to enable `show`ing values of your type."]
Show : A => trait (A -> Text)

[help "Display a value on the screen."]
show :: A where (Show A) => A -> ()
show : x -> external "runtime" "display" (Show x)

instance (Show Text) : it
instance (Show Number) : n -> external "runtime" "number-to-text" n
instance (Show Integer) : n -> external "runtime" "integer-to-text" n
instance (Show Natural) : n -> external "runtime" "natural-to-text" n
instance (Show Byte) : n -> external "runtime" "byte-to-text" n
instance (Show Signed) : n -> external "runtime" "signed-to-text" n
instance (Show Unsigned) : n -> external "runtime" "unsigned-to-text" n
instance (Show Float) : n -> external "runtime" "float-to-text" n
instance (Show Double) : n -> external "runtime" "double-to-text" n
instance (Show Boolean) : x? -> if x? "True" "False"
instance (Show Index-Error) : just "invalid index"
instance (Show None-Error) : just "found `None`"

Value where (Show Value) => instance (Show (Maybe Value)) : x? -> when x? {
	Some x -> format "Some _" (Show x)
	None -> "None"
}

Success Failure where (Show Success) (Show Failure) =>
	instance (Show (Result Success Failure)) :
		x? -> when x? {
		OK x -> format "OK _" (Show x)
		Error e -> format "Error _" (Show e)
	}

Success Failure where (Show Failure) =>
	instance (Expect (Result Success Failure) Success) :
		message -> x? -> when x? {
			OK x -> x
			Error e -> crash (format "_: _" message e)
		}

[help "Implement the `read` function."]
Read : A => trait (Text -> Maybe A)

[help "Parse `Text` into a value."]
read :: A where (Read A) => Text -> Maybe A
read : Read

instance (Read Text) : Some
instance (Read Number) : text -> external "runtime" "text-to-number" text
instance (Read Integer) : text -> external "runtime" "text-to-integer" text
instance (Read Natural) : text -> external "runtime" "text-to-natural" text
instance (Read Byte) : text -> external "runtime" "text-to-byte" text
instance (Read Signed) : text -> external "runtime" "text-to-signed" text
instance (Read Unsigned) : text -> external "runtime" "text-to-unsigned" text
instance (Read Float) : text -> external "runtime" "text-to-float" text
instance (Read Double) : text -> external "runtime" "text-to-double" text

[help "Read a value from the user with the provided prompt."]
prompt :: A where (Read A) => Text -> A
prompt : prompt -> external "runtime" "prompt" prompt (Read :: (Text -> Maybe A))

[help "Ask the user to pick from a set of options."]
choice :: Element where (Show Element) => Text -> List Element -> Element
choice : prompt -> items -> {
	descriptions : items . transform (((item :: Element) -> Show item) :: (Element -> Text)) . collect :: List Text
	index : external "runtime" "choice" prompt descriptions
	items . nth index . unwrap
}
