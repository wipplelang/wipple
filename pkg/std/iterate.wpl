[[no-std]]

use "util.wpl"
use "fp.wpl"
use "mutable.wpl"
use "list.wpl"
use "loop.wpl"

[help "A container for an iterable value that produces the `next` value on demand."]
Iterator : Element => type {
	next :: () -> Maybe Element
}

[help "Implement this trait to enable traversing your collection type."]
Iterate : Container Element => trait (Container -> Iterator Element)

Element => instance (Iterate (Iterator Element) Element) : it

-- reduce :: Container Element Result where (Iterate Container Element) => (Element -> Result -> Result) -> Container -> Iterator Result
-- reduce : TODO

[help "Keep only the items that satisfy the provided condition."]
filter :: Container Element where (Iterate Container Element) => (Element -> Boolean) -> Container -> Iterator Element
filter : include? -> container -> {
	{ next } : Iterate container

	Iterator {
		next : () -> loop {
			when (next ()) {
				Some x ->
					if (include? x)
						(Break (Some x))
						(Continue :: Control-Flow (Maybe _))  -- type annotation is needed because of typechecker bug
				None -> Break None
			}
		}
	}
}

[help "Execute a function for each item in an iterator."]
each :: Container Element where (Iterate Container Element) => (Element -> ()) -> Container -> ()
each : f -> container -> {
	{ next } : Iterate container

	loop {
		when (next ()) {
			Some x -> {
				f x
				(Continue :: Control-Flow ()) -- type annotation is needed because of typechecker bug
			}
			None -> Break ()
		}
	}
}

[help "Implement this trait to enable reducing an iterator into your collection type."]
Collect : Element Container => trait (Iterator Element -> Container)

[help "Collect an iterator into a single value, like a `List`."]
collect :: Input Element Output where (Iterate Input Element) (Collect Element Output) => Input -> Output
collect : input -> Collect ((Iterate :: _ -> Iterator Element) input) -- FIXME: Remove type annotation once we have functional dependencies

Element => instance (Collect Element (List Element)) : { next } -> {
	list : mutable (list ())
	loop {
		when (next ()) {
			Some x -> {
				list . append! x
				Continue
			}
			None -> Break ()
		}
	}

	get list
}

A => instance (Iterate (List A) A) : list -> {
	index : mutable 0

	Iterator {
		next : () -> when (list . nth (get index)) {
			OK element -> {
				increment! index
				Some element
			}
			Error _ -> None
		}
	}
}
