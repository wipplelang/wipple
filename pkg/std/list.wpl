[[no-std]]

use "util.wpl"
use "fp.wpl"
use "mutable.wpl"

[help "Implementations for the `list` function."]
[on-unimplemented "lists may only be created from tuples where every value has the same type"]
Tuple-To-List : Tuple Element => trait (Tuple -> List Element)

tuple-to-list-impl : ty' tuple' ~>
	(ty' => instance (Tuple-To-List tuple' ty') : tuple -> external "runtime" "make-list" tuple)

tuple-to-list-impl A ()
tuple-to-list-impl A (A ,)
tuple-to-list-impl A (A , A)
tuple-to-list-impl A (A , A , A)
tuple-to-list-impl A (A , A , A , A)
tuple-to-list-impl A (A , A , A , A , A)
tuple-to-list-impl A (A , A , A , A , A , A)
tuple-to-list-impl A (A , A , A , A , A , A , A)
tuple-to-list-impl A (A , A , A , A , A , A , A , A)
tuple-to-list-impl A (A , A , A , A , A , A , A , A , A)
tuple-to-list-impl A (A , A , A , A , A , A , A , A , A , A)

[help "Create a `List` from a tuple."]
list :: Tuple Element where (Tuple-To-List Tuple Element) => Tuple -> List Element
list : Tuple-To-List

[help "Retrieve the first item inside a `List`, if it exists."]
first :: A => List A -> Maybe A
first : list -> external "runtime" "list-first" list

[help "Retrieve the last item inside a `List`, if it exists."]
last :: A => List A -> Maybe A
last : list -> external "runtime" "list-last" list

[help "Retrieve all but the last item inside a `List`."]
initial :: A => List A -> Maybe (List A)
initial : list -> external "runtime" "list-initial" list

[help "Retrieve all but the first item inside a `List`."]
tail :: A => List A -> Maybe (List A)
tail : list -> external "runtime" "list-tail" list

Index-Error : type

[help "Retrieve the value in a `List` at a particular index, if it exists."]
nth :: A => Natural -> List A -> Result A Index-Error
nth : index -> list -> external "runtime" "list-nth" list index

[help "Add a new item to the end of a `List`."]
append :: A => A -> List A -> List A
append : new -> list -> external "runtime" "list-append" list new

[help "Add a new item to the end of a `Mutable List`."]
append! :: A => A -> Mutable (List A) -> ()
append! : new -> list -> list . set! (list . get . append new)

[help "Insert a new item at a particular index inside a `List`."]
insert :: A => Natural -> A -> List A -> Result (List A) Index-Error
insert : index -> new -> list -> external "runtime" "list-insert" list index new

[help "Insert a new item at a particular index inside a `Mutable List`."]
insert! :: A => Natural -> A -> Mutable (List A) -> Result () Index-Error
insert! : index -> new -> list -> when (list . get . insert index new) {
	OK new-list -> {
		list . set! new-list
		OK ()
	}
	Error e -> Error e
}

[help "Remove the item at a particular index inside a `List`."]
remove :: A => Natural -> List A -> Result (List A) Index-Error
remove : index -> list -> external "runtime" "list-remove" list index

[help "Remove the item at a particular index inside a `Mutable List`."]
remove! :: Natural -> Mutable (List _) -> Result () Index-Error
remove! : index -> list -> when (list . get . remove index) {
	OK new-list -> {
		list . set! new-list
		OK ()
	}
	Error e -> Error e
}
