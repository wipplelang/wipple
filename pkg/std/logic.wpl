[[no-std]]

[language "boolean"]
[help "Represents either `True` or `False`."]
Boolean : type {
	False
	True
}

use Boolean

[keyword]
if : bool then else ~> when bool {
	False -> else
	True -> then
}

[keyword]
if? : bool then ~> if bool then ()

[help "Returns `False` if provided `True`, and vice versa."]
not :: Boolean -> Boolean
not : bool -> if bool False True

And : Left Right Output => trait (Left -> (() -> Right) -> Output)
and : conjunction operator (a b ~> And a (() -> b))
instance (And Boolean Boolean Boolean) : a? -> b? -> if a? (b? ()) False

Or : Left Right Output => trait (Left -> (() -> Right) -> Output)
-- `or` operator is defined in the compiler to accommodate `or` patterns
instance (Or Boolean Boolean Boolean) : a? -> b? -> if a? True (b? ())

[help "The result of a comparison between two values."]
Ordering : type {
	Less
	Equal
	Greater
}

[help "Implement this trait to enable comparisons between two values of your type."]
Order : A => trait (A -> A -> Ordering)

instance (Order Number) : a -> b -> external "runtime" "number-ordering" a b
instance (Order Integer) : a -> b -> external "runtime" "integer-ordering" a b
instance (Order Natural) : a -> b -> external "runtime" "natural-ordering" a b
instance (Order Byte) : a -> b -> external "runtime" "byte-ordering" a b
instance (Order Signed) : a -> b -> external "runtime" "signed-ordering" a b
instance (Order Unsigned) : a -> b -> external "runtime" "unsigned-ordering" a b
instance (Order Float) : a -> b -> external "runtime" "float-ordering" a b
instance (Order Double) : a -> b -> external "runtime" "double-ordering" a b

< : comparison operator (a b ~> when (Order a b) {
	Ordering Less -> True
	Ordering Equal -> False
	Ordering Greater -> False
})

> : comparison operator (a b ~> when (Order a b) {
	Ordering Less -> False
	Ordering Equal -> False
	Ordering Greater -> True
})

<= : comparison operator (a b ~> when (Order a b) {
	Ordering Less -> True
	Ordering Equal -> True
	Ordering Greater -> False
})

>= : comparison operator (a b ~> when (Order a b) {
	Ordering Less -> False
	Ordering Equal -> True
	Ordering Greater -> True
})

[help "Implement this trait to enable equality testing between two values of your type."]
Equal : A => trait (A -> A -> Boolean)
= : comparison operator (a b ~> Equal a b)
/= : comparison operator (a b ~> not (a = b))

instance (Equal Text) : a -> b -> external "runtime" "text-equality" a b
instance (Equal Number) : a -> b -> external "runtime" "number-equality" a b
instance (Equal Integer) : a -> b -> external "runtime" "integer-equality" a b
instance (Equal Natural) : a -> b -> external "runtime" "natural-equality" a b
instance (Equal Byte) : a -> b -> external "runtime" "byte-equality" a b
instance (Equal Signed) : a -> b -> external "runtime" "signed-equality" a b
instance (Equal Unsigned) : a -> b -> external "runtime" "unsigned-equality" a b
instance (Equal Float) : a -> b -> external "runtime" "float-equality" a b
instance (Equal Double) : a -> b -> external "runtime" "double-equality" a b
instance (Equal ()) : () -> () -> True
instance (Equal Boolean) : a? -> b? -> a? and b? or not a? and not b?

Numeric-Impl : N => type {
	zero :: N
}

Numeric : N => trait (Numeric-Impl N)

instance (Numeric Number) : Numeric-Impl {
	zero : 0
}

instance (Numeric Integer) : Numeric-Impl {
	zero : 0
}

instance (Numeric Natural) : Numeric-Impl {
	zero : 0
}

instance (Numeric Byte) : Numeric-Impl {
	zero : 0
}

instance (Numeric Signed) : Numeric-Impl {
	zero : 0
}

instance (Numeric Unsigned) : Numeric-Impl {
	zero : 0
}

instance (Numeric Float) : Numeric-Impl {
	zero : 0
}

instance (Numeric Double) : Numeric-Impl {
	zero : 0
}
