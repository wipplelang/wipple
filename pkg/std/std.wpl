[[no-std]]

use "fp.wpl"
use "logic.wpl"
use "output.wpl"
use "util.wpl"
use "math.wpl"
use "mutable.wpl"
use "loop.wpl"
use "list.wpl"
use "iterate.wpl"
use "range.wpl"
