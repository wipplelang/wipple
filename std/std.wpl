[[no-implicit-use]]
[[help-url "https://wipple.dev/doc/std.html"]]

use "async"
use "fp"
use "io"
use "list"
use "logic"
use "math"
use "mutable"
use "random"
use "range"
use "repeat"
use "sequence"
use "text"
use "ui"
use "util"
