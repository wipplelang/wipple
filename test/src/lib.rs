#![allow(missing_docs)]

use rstest::rstest;

#[rstest]
#[timeout(std::time::Duration::from_secs(10))]
#[tokio::test]
async fn tests(#[files("tests/**/*.test.wipple")] file: std::path::PathBuf) {
    use futures::{future, FutureExt};
    use serde::de::DeserializeOwned;
    use serde::Serialize;
    use std::{
        fs, io,
        path::Path,
        sync::{Arc, Mutex},
    };
    use wipple_driver::util::lazy_static::lazy_static;

    fn read_binary<T: DeserializeOwned>(path: impl AsRef<Path>) -> T {
        wipple_driver::util::read_binary(io::BufReader::new(fs::File::open(path).unwrap())).unwrap()
    }

    let test = async move {
        let file_name = file.file_name().unwrap().to_string_lossy().into_owned();

        let code = fs::read_to_string(&file).expect("failed to read file");

        let (should_compile, should_warn) = if code.starts_with("-- [should compile]") {
            (true, Some(false))
        } else if code.starts_with("-- [should warn]") {
            (true, Some(true))
        } else if code.starts_with("-- [should error]") {
            (false, None)
        } else {
            panic!(
                "expected test to begin with [should compile], [should warn], or [should error]"
            );
        };

        let file = wipple_driver::File {
            path: file.to_string_lossy().into_owned(),
            visible_path: wipple_driver::util::get_visible_path(&file),
            code,
        };

        let base_interface =
            read_binary::<wipple_driver::Interface>("../artifacts/base.wippleinterface");

        let base_library = read_binary::<wipple_driver::Library>("../artifacts/base.wipplelibrary");

        let result = wipple_driver::compile(vec![file], vec![base_interface.clone()]);

        let render = wipple_render::Render::new();
        render.update(
            vec![base_interface, result.interface],
            Some(result.library.clone()),
            None,
        );

        let compiled = Mutex::new(true);
        let compiled_with_warnings = Mutex::new(false);
        let mut rendered_diagnostics = result
            .diagnostics
            .into_iter()
            .map(|diagnostic| {
                let render = &render;
                let compiled = &compiled;
                let compiled_with_warnings = &compiled_with_warnings;

                let rendered_diagnostic = render
                    .render_diagnostic(&diagnostic)
                    .unwrap_or_else(|| panic!("could not render diagnostic: {diagnostic:#?}"));

                match rendered_diagnostic.severity {
                    wipple_render::RenderedDiagnosticSeverity::Error => {
                        *compiled.lock().unwrap() = false;
                    }
                    wipple_render::RenderedDiagnosticSeverity::Warning => {
                        *compiled_with_warnings.lock().unwrap() = true;
                    }
                }

                render.render_diagnostic_to_debug_string(&rendered_diagnostic)
            })
            .collect::<Vec<_>>();

        rendered_diagnostics.sort();
        rendered_diagnostics.dedup();

        let compiled = compiled.into_inner().unwrap();
        let compiled_with_warnings = compiled_with_warnings.into_inner().unwrap();

        assert_eq!(should_compile, compiled);

        if let Some(should_warn) = should_warn {
            assert_eq!(should_warn, compiled_with_warnings);
        }

        let output = if compiled {
            let executable =
                wipple_driver::link(vec![base_library, result.library]).expect("linking failed");

            #[derive(Clone)]
            struct Value;

            struct Runtime;

            lazy_static! {
                static ref RUNTIME: tokio::runtime::Runtime =
                    tokio::runtime::Builder::new_multi_thread()
                        .worker_threads(1)
                        .enable_all()
                        .build()
                        .unwrap();
            }

            impl wipple_interpreter::Runtime for Runtime {
                type Value = Value;
                type JoinHandle = tokio::task::JoinHandle<()>;

                fn run(
                    future: impl future::Future<Output = ()> + Send + 'static,
                ) -> Self::JoinHandle {
                    RUNTIME.spawn(future)
                }

                async fn from_value(
                    _value: wipple_interpreter::Value<Self>,
                    _task: &wipple_interpreter::TaskLocals<Self>,
                    _context: &wipple_interpreter::Context<Self>,
                ) -> Self::Value {
                    unimplemented!()
                }

                async fn to_value(_value: Self::Value) -> wipple_interpreter::Value<Self> {
                    unimplemented!()
                }

                async fn with_functions<T>(
                    _f: impl FnOnce(&mut Vec<wipple_interpreter::StoredFunction<Self>>) -> T + Send,
                ) -> T {
                    unimplemented!()
                }
            }

            let output: Arc<Mutex<Vec<String>>> = Default::default();

            let options = wipple_interpreter::Options::<Runtime>::with_io(wipple_interpreter::Io {
                display: Arc::new({
                    let output = output.clone();
                    move |message| {
                        let output = output.clone();
                        async move {
                            output.lock().unwrap().push(message);
                            Ok(())
                        }
                        .boxed()
                    }
                }),
                prompt: Arc::new(|_, _| unimplemented!()),
                choice: Arc::new(|_, _| unimplemented!()),
                ui: Arc::new(|_, _| unimplemented!()),
                sleep: Arc::new(|_| unimplemented!()),
            });

            if let Err(error) = wipple_interpreter::evaluate(executable, options).await {
                output.lock().unwrap().push(format!("error: {}\n", error.0));
            }

            let output = output.lock().unwrap();
            output.clone()
        } else {
            Vec::new()
        };

        #[derive(Default, Serialize)]
        struct Snapshot {
            diagnostics: Vec<String>,
            output: Vec<String>,
        }

        let snapshot = [rendered_diagnostics.join("\n"), output.join("\n")]
            .into_iter()
            .filter(|s| !s.is_empty())
            .collect::<Vec<_>>()
            .join("\n\n");

        insta::assert_snapshot!(file_name, snapshot);
    };

    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_path("../snapshots");
    settings.bind_async(test).await;
}
