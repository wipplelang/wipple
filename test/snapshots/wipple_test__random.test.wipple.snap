---
source: test/src/lib.rs
expression: snapshot
---
0
1
Some 1
None
